src/Orleans.Transactions/InClusterTM/TransactionManager.cs

```diff
@@ -0,0 +1,695 @@
+using Orleans.Transactions.Abstractions;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+using System.Threading;
+using Orleans.Runtime;
+
+namespace Orleans.Transactions
+{
+    internal class TransactionManager : ITransactionManager, IDisposable
+    {
+        private const int MaxCheckpointBatchSize = 200;
+        private readonly TransactionsConfiguration config;
+        private readonly TransactionLog transactionLog;
+        private readonly ActiveTransactionsTracker activeTransactionsTracker;
+
+        // Index of transactions by transactionId.
+        private readonly ConcurrentDictionary<long, Transaction> transactionsTable;
+
+        // Transactions that have validated dependencies.
+        private readonly ConcurrentQueue<Transaction> dependencyQueue;
+
+        // Transactions that are waiting on group commit.
+        private readonly ConcurrentQueue<Tuple<CommitRecord, Transaction>> groupCommitQueue;
+
+        // Queue of committed transactions in commit order.
+        private readonly ConcurrentQueue<Transaction> checkpointQueue;
+        private readonly Queue<Transaction> checkpointRetryQueue = new Queue<Transaction>(MaxCheckpointBatchSize);
+
+        private readonly InterlockedExchangeLock dependencyLock;
+        private readonly InterlockedExchangeLock commitLock;
+        private readonly InterlockedExchangeLock checkpointLock;
+        private readonly Dictionary<ITransactionalResource, long> resources;
+        private readonly List<Transaction> transactions;
+
+        private long checkpointedLSN;
+
+        private readonly Timer gcTimer;
+
+        protected readonly Logger Logger;
+
+        public TransactionManager(TransactionLog transactionLog, TransactionsConfiguration config, Factory<string, Logger> logFactory)
+        {
+            this.transactionLog = transactionLog;
+            this.config = config;
+            this.Logger = logFactory(nameof(TransactionManager));
+
+            activeTransactionsTracker = new ActiveTransactionsTracker(config, this.transactionLog, logFactory);
+
+            transactionsTable = new ConcurrentDictionary<long, Transaction>(2, 1000000);
+
+            dependencyQueue = new ConcurrentQueue<Transaction>();
+            groupCommitQueue = new ConcurrentQueue<Tuple<CommitRecord, Transaction>>();
+            checkpointQueue = new ConcurrentQueue<Transaction>();
+
+            this.dependencyLock = new InterlockedExchangeLock();
+            this.commitLock = new InterlockedExchangeLock();
+            this.checkpointLock = new InterlockedExchangeLock();
+            this.resources = new Dictionary<ITransactionalResource, long>();
+            this.transactions = new List<Transaction>();
+
+            checkpointedLSN = 0;
+
+            gcTimer = new Timer(GC, null, Timeout.Infinite, -1);
+        }
+
+        #region ITransactionManager
+
+        public async Task StartAsync()
+        {
+            await transactionLog.Initialize();
```

Comment:
If there's an exception during this process then we will be in an invalid state. I'm not sure what the best solution is to make this resilient, but maybe it ought to be retriable. Looking at the implementation of the Azure-backed tx log, I see that the `Initialize` method does not reset all state to a consistent point.

Reply:
The TM needs a lot more love, but for now I'll make the TM and Memory log transient instead of singletons.  That way a failure in the initialization will cause the grain activation to fail and the next attempt will start with a clean state.